File: clj/14.clj
(ns advent.2018.14)

(def TARGET (+ 846601 10))

(defn generate-recipes [val1 val2]
  (map #(Integer/parseInt (str %)) (seq (str (+ val1 val2)))))

(defn add-recipes [recipes new-recipes]
  (if (zero? (count new-recipes))
    recipes
    (recur (conj recipes (first new-recipes))
           (rest new-recipes))))

(time
(loop [elf1 0
       elf2 1
       recipes [3 7]]
  (let [elf1-value (nth recipes elf1)
        elf2-value (nth recipes elf2)
        new-recipes (generate-recipes elf1-value elf2-value)
        joined-recipes (add-recipes recipes new-recipes)
        recipe-count (count joined-recipes)]
    (if (<= TARGET recipe-count)
      (println (take-last 10 joined-recipes))
      (recur (mod (+ 1 elf1 elf1-value) recipe-count)
             (mod (+ 1 elf2 elf2-value) recipe-count)
             joined-recipes)))))

File: js/14.js
const start = new Date()
const TARGET = `846601`
const TARGET_COUNT = TARGET.length

function generateRecipes (val1, val2) {
    return `${val1 + val2}`.split('').map(char =>
        parseInt(char, 10))
}

let elf1 = 0,
    elf2 = 1,
    recipes = [3, 7]

while (true) {
    let elf1Value = recipes[ elf1 ],
        elf2Value = recipes[ elf2 ],
        newRecipes = generateRecipes( elf1Value, elf2Value )

    newRecipes.forEach( recipe => recipes.push(recipe) )

    elf1 = (elf1 + elf1Value + 1) % recipes.length
    elf2 = (elf2 + elf2Value + 1) % recipes.length

    let last = recipes.slice(-TARGET_COUNT).join('')
    let last2 = recipes.slice(-TARGET_COUNT - 1, -1).join('')
    
    if ( last == TARGET ) {
        console.log( recipes.length - TARGET_COUNT )
        break
    } else if ( last2 == TARGET  ) {
        console.log( recipes.length - TARGET_COUNT - 1 )
        break
    }
}

console.log('time: ', new Date() - start)